### ISSUE TITLE
Casting to Optional type is highlighted as incorrect

### ISSUE THREAD

**Issue by twolodzko** [microsoft/pyright#825]
**Describe the bug**
This is highlighted as `Pyright (reportGeneralTypeIssues)`:

```python
condition = cast(Optional[bool], condition)
```

This also happens for other, optional casts.

**Expected behavior**
Allow optional type casts.

**VS Code extension or command-line**
Using VS Code (1.47), ver 1.1.52 of the extension.

**Comment by erictraut** [comment]
Pyright is doing the correct thing here. 

The `builtins.pyi` type stub defines the `cast` method with the following two overloads:
```
@overload
def cast(typ: Type[_T], val: Any) -> _T: ...
@overload
def cast(typ: str, val: Any) -> Any: ...
```
In other words, the first argument must either be an actual class or a str. 

PEP 484 is very clear about the meaning of `Type`:
The value corresponding to Type[C] must be an actual class object that's a subtype of C, not a special form.

That means `Optional` should not match against `Type`.

You have a couple of options here. I'll list them in order of preference.
1. Annotate the symbol `conditional` so it is the correct type `Optional[bool]`. This eliminates the need to add a `cast`. I use casts only as a last resort. I can't tell from your example whether `condition` was a parameter or a variable assigned earlier in the code. If possible, use a variable type annotation as follows: `condition: Optional[bool]`.
2. Place the `Optional[bool]` in quotes.

**Comment by twolodzko** [comment]
Thanks, I know the workarounds. Should have checked PEP, but mypy does not flag this as incorrect, but if I follow what you say, it would be a bug on mypy's side?

**Comment by erictraut** [comment]
I would consider it a bug in mypy or in the typeshed `builtins.pyi` stub.

**Comment by erictraut** [comment]
It appears that the implementation of `cast` accepts `Optional[bool]` as the first argument, so that would seem to support the assertion that the `typing.pyi` stub has a bug. It should probably define a third overload to the the `cast` function:
```
def cast(typ: Any, val: Any) -> Any: ...
```

**Comment by twolodzko** [comment]
@erictraut the PEP also seems to say that

> Note that it is legal to use a union of classes as the parameter for Type[], [...]

where `Optional[T]` is an union type by definition and as mentioned in several places PEP, e.g.

> Type checkers should move towards requiring the optional type to be made explicit.

**Comment by erictraut** [comment]
Thanks @twolodzko for pointing that out. Based on that, Type[T] should allow a union of classes. However, I still don't think it should allow `None` or a union that includes `None` (i.e. `Optional[T]`). That would violate a bunch of assumptions made in the stdlib typeshed stubs where `Type[T]` is used.

**Comment by erictraut** [comment]
I've confirmed that Pyright already supports unions of classes for `Type[T]`. It doesn't support `None` or `Optional[T]`, and I don't think it should.

**Comment by hauntsaninja** [comment]
PEP 484 also says:
> When used in a type hint, the expression None is considered equivalent to type(None).

So I think it's reasonable to read:
> The value corresponding to Type[C] must be an actual class object that's a subtype of C, not a special form.

in the case of None as:
> The value corresponding to Type[None] (so `C = None = type(None)` ) must be an actual class object that's a subtype of type(None), not a special form.

Or for another substitution, `Type[Optional[T]] == Type[Union[T, None]] == Type[Union[T, NoneType]]` which does look like a union of classes... `inspect.isclass(type(None))` does return True.

> That would violate a bunch of assumptions made in the stdlib typeshed stubs where Type[T] is used.

I'm curious what assumptions you're thinking of here? Or maybe phrased differently, are there examples where mypy does the wrong thing?

Also, even if typeshed were to define the `(Any, Any) -> Any` overload, it wouldn't help pyright figure out the return type. This is surprising as a user, given that that's `cast`'s one job. There's a reasonable PEP 484 argument for special-casing `cast` here if you decide not to allow the more general `Type[T]`

> The cast(t, x) call tells the type checker that we are confident that the type of x is t

Btw, thank you for all your work! I've been enjoying Pylance :-)

**Comment by erictraut** [comment]
Consider the following code:

```
from typing import Type, TypeVar, Union, cast

_T = TypeVar("_T")

def my_func(a: Type[_T]) -> _T:
    return a()

my_func(None)
```

If I run this in the Python interpreter, I receive the error: 
```
Traceback (most recent call last):
  File "foo.py", line 10, in <module>
    foo(None)
  File "foo.py", line 7, in foo
    return a()
TypeError: 'NoneType' object is not callable
```

If I run mypy on this code, it emits the following error:
```
test.py:10: error: Argument 1 to "foo" has incompatible type "None"; expected "Type[<nothing>]"
Found 1 error in 1 file (checked 1 source file)
```

So this is evidence mypy doesn't accept None as a match for `Type[T]`. Plus, there are trivial cases where allowing this will result in uncaught bugs. It appears that mypy is ignoring the stub signature for `cast`, which I would consider a bug in mypy. 

Since `Optional[T]` is allowed by `cast`, then there's also a bug in the typeshed typing.pyi stub. It should include an overload that accepts `Any` and returns `Any`. Even better, it could accept a TypeVar and return the same TypeVar.

In any case, I think Pyright is doing the right thing here. I'm going to mark this "as designed" and close the issue. Feel free to file bugs against the mypy and typeshed repos.

**Comment by twolodzko** [comment]
@erictraut Your example is incorrect, try:

```python
my_func(type(None))
```

It doesn't fail, since you pass the `NoneType` type, rather then `None` itself. `Optional` is an union of a type and a `NoneType`.

**Comment by erictraut** [comment]
That's a good point. Reopening for another look.

**Comment by erictraut** [comment]
This will be fixed in the next version of Pyright.

**Comment by twolodzko** [comment]
Sounds great, thanks!

**Comment by erictraut** [comment]
This is addressed in Pyright version 1.1.57, which I just published. It will also be included in the next version of Pylance.
